### ISSUE TITLE
Work on BE of Labeling Feature

### ISSUE THREAD

**Issue by RDmitchell** [SEED-platform/seed#494]
Work with Daniel on the back end of the labeling feature. See #493 for the spec. 

@pipermerriam -- I am assigning this to you. 

Thanks !

**Comment by pipermerriam** [comment]
@RDmitchell there has been talk of using Django Rest Framework in this project.  I think I can do this feature quicker using that framework as well as establishing an example implementation of how to do API endpoints using DRF.  Thoughts?

@danielmcquillen if I get a green light on this, I'll probably make some modifications to the format of the data that goes into the API endpoints and PR the relevant FE changes into your branch.  Is that ok with you?

**Comment by rebrownlbnl** [comment]
I thought there was some discussion about not wanting to piecemeal switch over to DRF, that it should be done all at once?
I like the idea of doing an example implementation for one endpoint(s) to see how it goes, if it doesn't introduce more inconsistencies between endpoints.
@nllong were you cautioning against doing the incremental switch?

**Comment by RDmitchell** [comment]
or possible @nicholasserra was cautioning about the incremental approach?

**Comment by pipermerriam** [comment]
Interesting.  I would highly recommend an iterative approach.  Huge refactors like that are problematic in my experience.  Since these endpoints are *new* this seems like a nice place to start if the developers are ok with this.

For reference, I have  a **ton** of DRF experience and would plan on doing this in the simplest way possible so that it's an easy reference for other endpoints if we chose to continue that direction.

**Comment by rebrownlbnl** [comment]
Could you post some more information about how you would implement this in DRF?
The potential downsides seem to be that 
1) the affected endpoints could behave differently than other "standard" SEED endpoints, causing confusion, or
2) other endpoints break somehow from this change.
Any other risks to be aware of?

**Comment by pipermerriam** [comment]
I would like to make the new endpoints in a more REST-ful manner.  Use of `POST/PUT/GET/DELETE` http verbs for Create/Update/Retrieve/Delete operations.  Similarly, I'd like to use proper response status codes for responses rather than a `{"status": "success"}` json response and always returning a 200 status code.  These changes would require very minor modifications to the Frontend code that is already being worked on for this feature.

All of the current mechanisms for authentication and authorization would remain the same.

None of this would have any effect on existing endpoints.

The only risks are whether this project wants to transition over to DRF.  If the answer is *yes* or *maybe* then this is a low risk way to see what that actually would look like.  If later the migration to DRF never gets prioritized, then this project will have one small set of endpoints that don't conform to the remainder of the project, or these endpoints will have to be redone in the old style or some other style.  All of these risks seems relatively low to me.

**Comment by danielmcquillen** [comment]
If we are going to do a test, I'd agree with Piper that this seems like a nicely decoupled place to do it. 

Piper, would you apply REST principles to just the Label CRUD or would you also modify the method that applies/removes labels to buildings (since I'd imagine that's just a bit farther reach)?

And would using DRF require a change in BE method documentation format (i.e. structuring comments in some way to allow for automated creation of API sandbox)?

Perhaps a simple test like this might pay dividends in that -- if we find something amiss we didn't predict -- it prevents a complete update of all APIs and it only affects a less-that-critical feature.

**Comment by pipermerriam** [comment]
I'd plan on using DRF for both, but the multi-building-update would still look largely the same but with DRF wrapping the functionality.  This gives two nice examples for people to reference.  One that leverages the built-in CRUD capabilities, and one that shows how to implement a very custom API endpoint.

**Comment by RDmitchell** [comment]
I think the concern on the call was moving to REST for the existing APIs. Since this is a new section, it seems to me that it is worth doing as a test of the concept and to see if there are any unforseen ramifications, as Daniel mentions. I would vote for doing it. Rich, you should weigh in based on this new information.

**Comment by nicholasserra** [comment]
In my opinion, correct request methods (POST, PUT, etc) and correct status
codes are just called good programming. Not really any choice there.

So i'm +1 with doing the correct requests/responses for these new endpoints.

I'm +0 on DRF on these. I don't like things piece by piece when it comes to
api stuff, especially since DRF seems like it's more for building an entire
api resource. But if you (piper) think that shipping these one off
endpoints with DRF isn't a big deal, then I'll go with your experience on
it.

**Comment by rebrownlbnl** [comment]
Thanks everyone for weighing in.

Overall this seems pretty low risk, and I do see us moving to DRF in the
future so this will be a good test.
I hear Nick Serra about not mixing API styles, but I think the benefits
outweigh the drawbacks.
Let's go with Piper's approach.

**Comment by mmclark** [comment]
I would agree.  There was some talk in the last developer's call about possibly using DRF for a v2 of the API (in parallel with the existing v1).  It seems like this is a good first step towards that.

**Comment by pipermerriam** [comment]
@StephenCzarnecki According to an email I was forwarded, you may have started some work on migrations for labels.  Is there a branch where I can view this work?

**Comment by rebrownlbnl** [comment]
Let me know if this is overly ambitious, but I think it's worth writing down a few guidelines here for implementing the revised API:

1) use only HTTP verbs wherever possible (`POST/PUT/GET/DELETE`), 
2) endpoint names should be plural nouns,
3) use proper HTTP response codes (can we define what is proper behavior in some ambiguous cases, like when you try to create a resource that already exists?),
4) versioning: do we put this new endpoint(s) under a `/v2/` branch and keep the existing endpoints in parallel?
5) query parameters: do we need to specify a standard format for filtering, sorting, paging, etc.?

If this is best left for a future release, let's move this discussion to a separate thread and work on a more comprehensive set of design guidelines.

**Comment by pipermerriam** [comment]
I'd like to focus on the implementation first to be sure that deadlines don't get missed, and then backfill this spec information.  Is that ok?

**Comment by pipermerriam** [comment]
If there are no objections, I would like to go ahead and namespace these new urls under `/v2/`.

**Comment by StephenCzarnecki** [comment]
Hi

I replied in another thread, the one with subject "SEED -- Review (and
possibly implementation) of Label feature backend proposal"

**Comment by rebrownlbnl** [comment]
I think that's fine to implement this as a "proof of concept" for new APIs, and complete the spec later. In that case, let's save `v2` for our complete future set of APIs, and let's version this interim API `/v1/' (or `/v1.3/` if we want to be consistent with the SEED release numbering, since we plan to increment to v1.3 with the December release).

**Comment by pipermerriam** [comment]
@rebrownlbnl  Rather than namespace these all by themselves, I think I'd opt towards just creating them in the same style as the current urls, and they can be migrated to the `/v2/` namespace when this approach gets formalized.  How does that sound?

**Comment by rebrownlbnl** [comment]
Sounds like a good approach.

**Comment by pipermerriam** [comment]
Just an update for anyone who's interested.

I've got all of the CRUD operations working correctly and merged into @danielmcquillen 's branch.  Things are going smoothly thus far.  If you'd like to take a look at the implementation you can see it in this branch https://github.com/SEED-platform/seed/tree/434-develop-labels

**Comment by rebrownlbnl** [comment]
One thought about testing: @bravacheLBL has an API testing script he maintains for hitting the API endpoints over http. He'll need to update the script to handle your new/modified endpoints.

**Comment by pipermerriam** [comment]
@rebrownlbnl DRF comes with it's own testing helpers that make testing API endpoints pretty easy.

**Comment by pipermerriam** [comment]
Just a quick update.

* This is coming along swimmingly thus far.

* [x] admin CRUD management of labels
* [x] filtering by labels in the building list.
* [x] bulk updating of labels in the buildings list.
* [ ] editing a single building's labels on the building information screen.

There is still work to be done with getting tests for this new functionality.  On track to have it code-complete by early tomorrow if not EOD today.

@danielmcquillen , can you speak to the state of the FE for this?  Seems largely complete with the exception of the building information page.

**Comment by RDmitchell** [comment]
How do you make checked and unchecked boxes for bullets ?? very nice !

And thanks for the update !

**Comment by pipermerriam** [comment]
```
- [ ] this will be unchecked
- [x] this will be checked
```

**Comment by RDmitchell** [comment]
cool ! thx ...

**Comment by rebrownlbnl** [comment]
re. testing, can @pipermerriam point us to the docs on DRF test helpers, or write up a little summary to help @bravachelbl extend the API tests? Thanks

**Comment by pipermerriam** [comment]
@rebrownlbnl the documentation for the DRF testing helpers can be found [here](http://www.django-rest-framework.org/api-guide/testing/)

Once I have some API tests to show off I'll see about a brief writup on best practices.

**Comment by RDmitchell** [comment]
We will move implementing the management of labels from the Building Detail screen to the Feb 2016 release. See issue #515.

**Comment by pipermerriam** [comment]
@RDmitchell 

@danielmcquillen just brought up another potentially missing part of this feature.  There is a data migration that will populate all organizations with the appropriate default labels.  However, there is no mechanism to create these default labels for *new* organizations.

Would it be good for newly created orgs to have the default label set created for them?

**Comment by RDmitchell** [comment]
Yes. I thought we were going to use the labels in this doc as the default set of labels. So I guess this would apply to new orgs.

https://docs.google.com/a/lbl.gov/document/d/14tvpbVFQXDihItRvrC4WDCnNqiMRbFrOtNbpWIq_Vys/edit?usp=sharing

**Comment by RDmitchell** [comment]
I am closing this as we have implemented the first phase of labeling and now can open separate issues for bugs and enhancements. Thanks Piper !
